test/functional/lua/snippet_spec.lua

```diff
@@ -0,0 +1,259 @@
+      {
+        base_indent = [[  ]],
+        indent_setting = [[
+          vim.o.expandtab = true
+          vim.o.shiftwidth = 2
+        ]],
+        expects = {
+          '  class ClassName {',
+          '    public ClassName() {',
+          '      ',
+          '    }',
+          '  }',
+          '  ',
+        }
+      }, {
+        base_indent = [[  ]],
+        indent_setting = [[
+          vim.o.expandtab = true
+          vim.o.shiftwidth = 0
+          vim.o.tabstop = 2
+        ]],
+        expects = {
+          '  class ClassName {',
+          '    public ClassName() {',
+          '      ',
+          '    }',
+          '  }',
+          '  ',
+        }
+      }, {
+        base_indent = [[<Tab>]],
+        indent_setting = [[
+          vim.o.expandtab = false
+        ]],
+        expects = {
+          '\tclass ClassName {',
+          '\t\tpublic ClassName() {',
+          '\t\t\t',
+          '\t\t}',
+          '\t}',
+          '\t',
+        }
+      }
+    }) do
+      clear()
+      exec_lua(case.indent_setting)
+      feed('i' .. case.base_indent)
+      exec_lua('vim.snippet.expand(...)', snippet)
+      eq(case.expects, helpers.buf_lines(0))
+    end
+  end)
+
+  it('should able to jump to all placeholders', function()
+    exec_lua('vim.snippet.expand(...)', table.concat({
+      'class ${1:ClassName} {',
+      '\tpublic $1($2) {',
+      '\t\t${3}',
+      '\t}',
+      '}',
+      ''
+    }, '\n'))
+    local cases = {
+      { m = 's', s = { 0, 6 }, e = { 0, 14 }, },
+      { m = 'i', s = { 1, 18 }, e = { 1, 18 }, },
+      { m = 'i', s = { 2, 2 }, e = { 2, 2 }, },
+      { m = 'i', s = { 5, 0 }, e = { 5, 0 }, },
+    }
+    for i = 1, #cases do
+      eq(cases[i], exec_lua([[return get_state()]]))
+      eq(i ~= #cases, exec_lua([[return vim.snippet.jumpable(vim.snippet.JumpDirection.Next)]]))
+      feed('<Tab>')
+    end
+    eq(cases[#cases], exec_lua([[return get_state()]]))
+    for i = #cases, 1, -1 do
+      eq(cases[i], exec_lua([[return get_state()]]))
+      eq(i ~= 1, exec_lua([[return vim.snippet.jumpable(vim.snippet.JumpDirection.Prev)]]))
+      feed('<S-Tab>')
+    end
+    eq(cases[1], exec_lua([[return get_state()]]))
+  end)
+
+  it('should sync same tabstop marks', function()
+    exec_lua('vim.snippet.expand(...)', table.concat({
+      'class ${1:ClassName} {',
+      '\tpublic $1($2) {',
+      '\t\t${3}',
+      '\t}',
+      '}',
+      ''
+    }, '\n'))
+    feed('ModifiedClassName<C-l>')
+    eq({
+      'class ModifiedClassName {',
+      '\tpublic ModifiedClassName() {',
+      '\t\t',
+      '\t}',
+      '}',
+      '',
+    }, helpers.buf_lines(0))
+  end)
+
+  -- it('should insert selected choice', function()
+  --   exec_lua('vim.snippet.expand(...)', table.concat({
+  --     'console.${1|log,info,warn,error|}($2);',
+  --   }, '\n'))
+  --   feed('3<CR>')
+  --   eq({
+  --     'console.warn();',
+  --   }, helpers.buf_lines(0))
+  -- end)
```

Comment:
How do I write test for this?

I can't manage `vim.schedule` and `feedkeys`...